### ISSUE TITLE
Feature Request - Encrypt resource.cfg or use another method to store login credentials securely

### ISSUE THREAD

**Issue by ssaxnagios** [NagiosEnterprises/nagioscore#367]
Encrypt resource.cfg or use another method to store login credentials securely.

It would be nice if there was a credentials manager for Core but we should at least not store them in cleartext for people who require more security.

**Comment by ericloyd** [comment]
Pna jr whfg ebg13 gurz?

**Comment by box293** [comment]
I like the idea of a credentials manager as apposed to changing how resource.cfg works. resource.cfg provides capabilities other than storing passwords, $USER1$=/usr/local/nagios/libexec

**Comment by ssaxnagios** [comment]
ericloyd, Vg'f hc gb gur qrif.

**Comment by jfrickson** [comment]
I wonder if B64 would be enough.

**Comment by ericloyd** [comment]
Any ecnonidg taht lveaes it in piatlxent wtiuhot a slat for eipnocrytn is dmooed to fail as far as sceriuty is ccrnoednd.

**Comment by amprantino** [comment]
Yes, that would be great!

**Comment by lancefogle** [comment]
@jfrickson Base64 is not encryption at all but is rather an encoding and is thus reversible in a matter of seconds.

I personally think @box293's idea to provide a custom credentials manager is a winner in this regard and that way you don't have to touch the resources.cfg file in any way and it can continue to be used for storing commands and other semi-sensitive data in variables outside the CGI scripts.

Passwords alone need better protection on systems and a credential manager also gives you the ability to later expand it to include different capabilities such as utilizing oauth endpoints for authenticating a plugin and checking a service or something **snazzy** like that. Room to grow and all that.

1. A credential manager that stored creds using one-way hashes is the industry "best" route for storing credentials but unfortunately that doesn't work for credentials used in remote authentication because the target system doesn't expect hashes but rather the plain-text password
2. Because of **<1>** you are forced to store the creds on the Nagios system with reversible encryption at best so that Nagios can decrypt the creds before sending them to the target system

So basically you would need a system that stores the credentials with reversible encryption secured likely with a private key that would need to be created and stored with proper file permissions to prevent unauthorized users from being able to read it.

The credential manager can generate a key on first use and store it with read rights for only the user Nagios runs as and store it in the home directory of that user? Or maybe in the Nagios ```/etc/``` directory for those users who don't want their application user accounts like ```nagios``` to have home directories?
- Passwords are not in plain-text and can be encrypted using strong algorithms
- Secures the private key from prying eyes
- Allows storing creds in a reversible fashion for authenticating to remote systems
- Could potentially support any encryption library out there as it could just pick up what is available on the system at run-time instead of compile-time or even support a run-time config file for customizing what algorithm or library you want to use so companies can go with whatever encryption they want to meet their security policies

**Comment by ericloyd** [comment]
Yes, a credential manager is the way to go.  I think every other reply here was mostly joking.  I know I was.  However, a credential manager increases complexity but, as you mentioned, it has its own complexities.  

And, just to mess with his head a bit, it wasn't @box293 who suggested it, it was @ssaxnagios  :-)

**Comment by ssaxnagios** [comment]
@ericloyd You da real MVP!

**Comment by ericloyd** [comment]
Lbh pna cnl zr va orref.

**Comment by lancefogle** [comment]
Haha fair enough

**Comment by hedenface** [comment]
This is a **great** idea. I'm tentatively scheduling it for core 5, and we'll discuss it in the next meeting.

**Comment by ygt-francoisc** [comment]
@hedenface do you have any updates on this ? Is a credential management feature somewhere in the 
 pipeline ?

**Comment by ericloyd** [comment]
We've started encoding all of our sensitive data in Pig Latin that is then rot13 the result:
```
Rgunl nffjbeqcnl vflnl Hsslsynl Haalonl
```

Note: I'm jkonig.  We stlil use palinetxt lkie nraoml haumns.

**Comment by iam333** [comment]
Hi!
Thanks in advance!

I´m looking for encript some passwords used in Nagios commands. I´ve tried to use resources.cfg, but....there is no the solution, if there is a problem with host , the service asociated to this host shows the command with the password and this is a great security hole...

is there any way to encryt this? 

Thanks

**Comment by ericloyd** [comment]
Unfortunately, unless you write your own password manager, and use a wrapper script for all your plugins to decode the password(s) before use, there is no way to manage encrypted passwords within Nagios at this time.  See earlier notes in this thread about the discussion of a credentials manager.

**Comment by iam333** [comment]
Hi! Thanks!

Is there any development line to include it in future release or versions?
I think this too basic in systems administration areas... to maintain the security.

thank you

**Comment by sawolf** [comment]
Hi @iam333,

Currently, it's in on the list for the Core 5 milestone. Unfortunately, I'm not able to give you more detail than that.

**Comment by iam333** [comment]
Hi!!
Thanks for information!
It´s great notice! It´s so important in my opinion.
